File: 0140-word-break-ii/0140-word-break-ii.java
class Solution {
    public List<String> wordBreak(String s, List<String> wordDict) {
        Set<String> set = new HashSet<>();
        for (String word : wordDict) {
            set.add(word);
        }
        
        
        List<String> res = new ArrayList<>();
        List<String> path = new ArrayList<>();
        backtrack(s, 0, path, set, res);
        return res;
    }
    
    private void backtrack(String s, int start, List<String> path, Set<String> set, List<String> res) {
        if (start == s.length()) {
            res.add(String.join(" ", path));
            return;
        }
        
        for (int i = start; i < s.length(); i++) {
            if (set.contains(s.substring(start, i + 1))) {
                path.add(s.substring(start, i + 1));
                backtrack(s, i + 1, path, set, res);
                path.remove(path.size() - 1);
            }
            
        }
    }
}

File: 0140-word-break-ii/0140-word-break-ii.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        wordSet = set(wordDict) 
        memo = {} 

        def backtrack(start):
            if start in memo:
                return memo[start]
            
            if start == len(s):
                return [""]
            
            result = []
            for end in range(start + 1, len(s) + 1):
                word = s[start:end]
                if word in wordSet:
                    rest_sentences = backtrack(end)  # Recursive call
                    for sentence in rest_sentences:
                        result.append(word + (" " + sentence if sentence else ""))

            memo[start] = result  # Store result in memo
            return result

        return backtrack(0)